### PR TITLE
WIP: Reading capabilities from disk

### PR DESCRIPTION
**Pull request by thenor**
Parses capability from Service Manifest files and returns GatewayConfigurations.

### CODE REVIEW

File: agent/src/capability/configstore.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ConfigStore::ConfigStore()
+{
+}
+
+ConfigStore::~ConfigStore()
+{
+}
+
+    static std::vector<GatewayConfiguration> capFromFile(std::string capID, std::string &configFilePath);
+    static std::vector<GatewayConfiguration> capFromDir(std::string capID, std::string &configDirPath);
```
Comment (tobsan) [inline agent/src/capability/configstore.cpp]: Indentation error?

**Review by tobsan — COMMENTED** [review]
From the looks of this it looks like we'll be reading in capabilities from file on demand. Am I reading it correct then? If so, that should be changed to reading all of them at once and then do a lookup in some data structure on demand.

File: agent/src/capability/configstore.cpp (hunk 1 above)
Comment (thenor) [reply agent/src/capability/configstore.cpp]: Copy paste error from the h file. I saw this myself. I also saw that my cpp and h files are not aligned at all. Sorry about that.

File: agent/src/capability/configstore.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * @brief Returns all Gateways for a certain Capability.
+     */
+    static std::vector<GatewayConfiguration> getGatewaysByCap(std::string capID, std::vector<Capability> caps);
+
+    /**
+     * @brief Populates a vector with Gateway objects and returns true on
+     * 	success and false on failure.
+     */
+    static bool parseGatewayData(const pelagicore::JSonElement &gatewayData, std::vector<Gateway> &gateways, std::string *error);
```
Comment (thenor) [inline agent/src/capability/configstore.h]: "const json_t *element" ?? Not "const pelagicore::JSonElement &gatewayData" ??

File: agent/src/capability/configstore.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+{
+  
+}
+
+ConfigStore::~ConfigStore()
+{
+}
+
+std::vector<GatewayConfiguration> softwarecontainer::capFromFile(std::string capID,
+								 std::string &filePath)
```
Comment (DunderRoffe) [inline agent/src/capability/configstore.cpp:33]: fix indentaion

File: agent/src/capability/configstore.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+        caps.insert(caps.end(), configCaps.begin(), configCaps.end());
+    }
+
+    return caps;
+}
+*/
+
+std::vector<GatewayConfiguration> softwarecontainer::getGatewaysByCap(std::string capID,
+								      const std::string &filePath) {
```
Comment (DunderRoffe) [inline agent/src/capability/configstore.cpp:76]: align `const std::string &filePath` with `std::string capID`

File: agent/src/capability/configstore.cpp (hunk 3 above)
Comment (DunderRoffe) [reply agent/src/capability/configstore.cpp:33]: Since `filePath` is const in getGatewayCap(...), maybe `filePath` could be const here as well?

**Review by DunderRoffe — COMMENTED** [review]
Seems good overall, but the indentation is a bit off.

File: agent/src/capability/configstore.cpp (hunk 4 above)
Comment (DunderRoffe) [reply agent/src/capability/configstore.cpp:76]: This applies generally in the file

File: agent/src/capability/configstore.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+//#include "SoftwareContainerAgent_dbuscpp_adaptor.h"
+#include "softwarecontainer-common.h"
+
+#include <jsonparser.h>
+#include "commandjob.h"
+
+
+namespace softwarecontainer {
+
+class ConfigStore
```
Comment (frznlogic) [inline agent/src/capability/configstore.h:46]: I'd love some documentation on what ConfigStore is.

File: agent/src/capability/configstore.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+     *                                                                                                                          
+     */
+    static std::vector<GatewayConfiguration> capFromDir(std::string capID, std::string &configDirPath);
+
+private:
+
+    /**
+     * @brief Returns all Gateways for a certain Capability.
+     */
+    static std::vector<GatewayConfiguration> getGatewaysByCap(std::string capID, const std::string &filePath);
```
Comment (frznlogic) [inline agent/src/capability/configstore.h:80]: Some of these declarations wind up really long, not sure if what we say about using namespace, and also maybe declare a new datatype for vector<GatewayConfiguration>. Just a thought.

File: agent/src/capability/configstore.cpp (hunk 4 above)
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:76]: the { of functions are supposed to be on a new line?

File: agent/src/capability/configstore.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+		    auto gatewaysElement = child.getField("gateways");
+		    if (arrayElement.isArray()) {
+
+		        std::string strError;
+			if (!parseGatewayData(gatewaysElement, gateways, &strError)) {
+			  log_debug() << "Error in " << pathString << ": " << strError;
+			  gateways.clear();
+			}
+			else {
+			  GatewayConfiguration gw = GatewayConfiguration(gatewayName, gwConf);
```
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:104]: There's a lot of nesting going on here... can the function be broken down?

File: agent/src/capability/configstore.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            if (pelagicore::has_suffix(filename, ".json"))
+                configFiles.push_back(filename);
+        }
+        closedir(dirFile);
+    }
+
+    return configFiles;
+}
+
+bool softwarecontainer::hasEnding(std::string const &fullString, std::string const &ending)
```
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:167]: These  functions are declared in the softwarecontainer::ConfigStore in the header file, but here they are in softwarecontaner:: only?

File: agent/src/capability/configstore.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+{
+    std::vector<GatewayConfiguration> GWconfs;
+
+    if (!configFilePath.empty()) {
+        GWconfs = getGatewaysByCap(capID, filePath);
+    }
+
+    return GWconfs;
+}
+/*
```
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:43]: Why is this commented out?

File: agent/src/capability/configstore.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+    DIR *dirFile = opendir(configPath.c_str());
+    if (dirFile == nullptr) {
+        log_debug() << "Could not open directory: " << configPath << "\n";
+    } else {
+        struct dirent *hFile;
+        while ((hFile = readdir(dirFile)) != nullptr) {
+            std::string filename(hFile->d_name);
+
+            if (pelagicore::has_suffix(filename, ".json"))
```
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:158]: No if without {}

File: agent/src/capability/configstore.cpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            for (auto& child : children) {
+
+                std::string capName;
+                child.read("name", capName);
+
+		// TODO (thenor): Only push gateways for capID
+		if (capName == capID) {
+		    auto gatewaysElement = child.getField("gateways");
+		    if (arrayElement.isArray()) {
+
```
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:97]: Too many empty lines.

File: agent/src/capability/configstore.cpp [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+		        std::string strError;
+			if (!parseGatewayData(gatewaysElement, gateways, &strError)) {
+			  log_debug() << "Error in " << pathString << ": " << strError;
```
Comment (frznlogic) [inline agent/src/capability/configstore.cpp:100]: Indentation in general is wrong. (2 spaces only here)

**Review by frznlogic — CHANGES_REQUESTED** [review]
Looks like you are started. Some minor comments here and there. I think the getGatewaysByCap could use some breaking apart, possibly some generalization as well. Perhaps something in the jsonparser?

**Comment by thenor** [comment]
Thanks for all the comments. I will go through them, but some might be inapplicable in the new pull request. Stay tuned!

**Pull request closed without merging**